libraries/llvm/io.c

```diff
@@ -21,4 +25,541 @@ void c_io_println_String(String text) {
     putchar('\n');
 }
 
+
+// Promises
+// --------
+
+typedef enum { UNRESOLVED, RESOLVED, AWAITED } promise_state_t;
+
+typedef struct Listeners {
+    struct Neg listener;
+    struct Listeners* next;
+} Listeners;
+
+typedef struct {
+    uint64_t rc;
+    void* eraser;
+    promise_state_t state;
+    // state of {
+    //   case UNRESOLVED => NULL
+    //   case RESOLVED   => Pos (the result)
+    //   case AWAITED    => Nonempty list of listeners
+    // }
+    union {
+        struct Pos pos;
+        Listeners* listeners;
+    } payload;
+} Promise;
```

Comment:
@phischu can we define all structs (Neg, Pos, etc.) like this, so that we do not have to write `struct Pos` everywhere?